File: llm.py
import os
import time
import json
import requests
from abc import ABC, abstractmethod
from typing import List, Dict, Any, Optional

class ILLMProvider(ABC):
    @abstractmethod
    def generate(self, system_prompt: str, user_prompt: str) -> str:
        pass

class FakeLLM(ILLMProvider):
    def __init__(self):
        self.responses: List[str] = []
        self.requests: List[Dict[str, str]] = []

    def generate(self, system_prompt: str, user_prompt: str) -> str:
        self.requests.append({"system_prompt": system_prompt, "user_prompt": user_prompt})
        if self.responses:
            return self.responses.pop(0)
        return "Default fake response"

class GeminiLLM(ILLMProvider):
    def __init__(self, api_key: Optional[str] = None):
        self.api_key = api_key or os.environ.get("GEMINI_API_KEY", "")
        if not self.api_key:
            raise ValueError("API Key is missing")

    def generate(self, system_prompt: str, user_prompt: str) -> str:
        if not self.api_key:
            raise ValueError("API Key is missing")

        url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key={self.api_key}"
        headers = {"Content-Type": "application/json"}
        payload = {
            "system_instruction": {
                "parts": [{"text": system_prompt}]
            },
            "contents": [{
                "role": "user",
                "parts": [{"text": user_prompt}]
            }]
        }

        max_retries = 4
        last_err = None

        for attempt in range(1, max_retries + 1):
            try:
                res = requests.post(url, headers=headers, json=payload, timeout=30)
                if res.status_code == 200:
                    data = res.json()
                    candidates = data.get("candidates", [])
                    if candidates and "content" in candidates[0]:
                        parts = candidates[0]["content"].get("parts", [])
                        if parts and "text" in parts[0]:
                            return parts[0]["text"]
                    return ""
                
                # Check for rate limits / quota limits
                if res.status_code in [429, 503]:
                    wait_time = 15 if res.status_code == 429 else (2 ** (attempt - 1) * 2)
                    time.sleep(wait_time)
                    continue

                res.raise_for_status()
            except Exception as e:
                last_err = e
                time.sleep(2)

        raise last_err or RuntimeError("LLM API call failed")

File: src/llm.ts
import { GoogleGenAI } from '@google/genai';

export interface LLMRequest {
  systemPrompt: string;
  userPrompt: string;
}

export interface ILLMProvider {
  generate(request: LLMRequest): Promise<string>;
}

export class FakeLLM implements ILLMProvider {
  public responses: string[] = [];
  public requests: LLMRequest[] = [];

  async generate(request: LLMRequest): Promise<string> {
    this.requests.push(request);
    const response = this.responses.shift() || 'Default fake response';
    return response;
  }
}

export class GeminiLLM implements ILLMProvider {
  private ai: GoogleGenAI;
  private apiKey: string;

  constructor() {
    this.apiKey = process.env.GEMINI_API_KEY || '';
    if (!this.apiKey && process.env.NODE_ENV !== 'test') {
      console.warn('GEMINI_API_KEY environment variable is not set.');
    }
    
    // Initialize the official Google GenAI SDK
    this.ai = new GoogleGenAI({ apiKey: this.apiKey });
  }

  async generate(request: LLMRequest): Promise<string> {
    if (!this.apiKey) {
      throw new Error('API Key is missing');
    }
    const MAX_RETRIES = 4;
    let lastError: Error = new Error('Unknown error');

    for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
      try {
        const response = await this.ai.models.generateContent({
          model: 'gemini-3.6-flash',
          contents: request.userPrompt,
          config: {
            systemInstruction: request.systemPrompt,
          }
        });
        return response.text || '';
      } catch (e: any) {
        lastError = new Error(`LLM API Error: ${e.message}`);
        const isRetryable = e.message?.includes('503') || 
                            e.message?.includes('UNAVAILABLE') || 
                            e.message?.includes('overloaded') || 
                            e.message?.includes('429') || 
                            e.message?.includes('RESOURCE_EXHAUSTED') || 
                            e.message?.includes('quota');
        if (isRetryable && attempt < MAX_RETRIES) {
          // If quota exceeded, wait a bit longer to reset the RPM bucket (e.g. 15s, 30s)
          const isQuota = e.message?.includes('429') || e.message?.includes('RESOURCE_EXHAUSTED');
          const waitMs = isQuota ? Math.max(15000, 10000 * attempt) : (2000 * Math.pow(2, attempt - 1));
          console.warn(`[LLM] Rate limit or service busy. Pausing ${waitMs / 1000}s before retry... (attempt ${attempt}/${MAX_RETRIES})`);
          await new Promise(r => setTimeout(r, waitMs));
        } else {
          break;
        }
      }
    }
    throw lastError;
  }
}
